Give "Spun Out" dynamic spin rate

sh0ckR6 opened the pull request:
Resolves #16839

Uses `Spinner.SpinsRequired` and `Spinner.Duration` to figure out required RPM for a perfect spin.
Another 1% is added to the RPM to account for any decimal loss during calculations. Without this, some spinners would just barely not complete.

sh0ckR6 committed:
Give Spun Out mod dynamic spin rate

sh0ckR6 committed:
Add missing parentheses to RPM calculation

bdach commented:
[Tests in `TestSceneOsuModSpunOut`](https://github.com/ppy/osu/runs/5172465211#r0s0) are failing and need attention. It may be that they are no longer valid with this change, but if that is the case, they need to be removed (after giving the proper justification for doing so).

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModSpunOut.cs:
```diff
@@ -45,7 +45,8 @@ private void onSpinnerUpdate(Drawable drawable)
             // for spinners we want the real (wall clock) elapsed time; to achieve that, unapply the clock rate locally here.
             double rateIndependentElapsedTime = spinner.Clock.ElapsedFrameTime / spinner.Clock.Rate;
-            spinner.RotationTracker.AddRotation(MathUtils.RadiansToDegrees((float)rateIndependentElapsedTime * 0.03f));
+            float rotationSpeed = (float)(spinner.HitObject.SpinsRequired / (spinner.HitObject.Duration / 1.01));
```
I think it's a bit weird how the 1.01 factor is applied to the duration rather than the speed. Would prefer to write

```suggestion
            float rotationSpeed = (float)(1.01 * spinner.HitObject.SpinsRequired / spinner.HitObject.Duration);
```

The result should be theoretically the same, but I feel the above is clearer.

Also an inline comment about what the 1.01 adjustment is doing would be a good idea.

sh0ckR6 replied:
Totally agree, it looks much nicer and makes more sense to apply the factor to the resulting speed, not the duration. I've updated the PR with this and an inline comment.

sh0ckR6 committed:
Update RPM calculation for readability

Multiply the 1.01 factor to the resulting RPM, not to the duration.

sh0ckR6 commented:
I've looked into the test failure, and it looks like `SpinnerSpmCalculator` just gives up for some reason during testing. In normal gameplay, it calculates SPM just fine, but in testing I keep getting 0.
##### Testing
![image](https://user-images.githubusercontent.com/68568464/153771867-7c526c5c-9429-4930-a365-ed89c7497b76.png)

##### Gameplay
![image](https://user-images.githubusercontent.com/68568464/153771884-ca9eb3bd-b242-410c-9d27-4b02ed2e44f5.png)

I don't think this test is fully required as the Spun Out mod does a perfectly good job handling rate changes, but rather a new test that ensures that the spinner completes with no bonus spins. This would check the SPM by nature of the test and also ensures that the mod isn't doing anything it's not supposed to on both sides (bonus spins, 100/50/miss).

peppy commented:
Generally we don't see a failing test due to some change and say "ah that test isn't required so she'll be right", but instead try to understand why it broke (it may mean there are unintended side-effects of the change at hand). If the spin rate was reporting correctly before but no longer is then there's a reason for this.

You mention adding a new test, which is 100% required (it should fail without the changes proposed in this PR). Are you planning on looking into that?

sh0ckR6 commented:
> If the spin rate was reporting correctly before but no longer is then there's a reason for this.

Makes sense. I'll keep digging into this to figure out why this isn't working as expected

> You mention adding a new test, which is 100% required (it should fail without the changes proposed in this PR). Are you planning on looking into that?

I do plan on working on a test for this PR. Not too familiar with o!f's testing framework, so it may take a bit longer than I'm hoping for.

I'll probably hold off on pushing more commits until I get at least the broken test passing to prevent queueing builds that I know will fail.

peppy commented:
To give you an initial direction: your assumption above that the counter isn't working is not correct. I think you were seeing the zeroes at the initial portion of the spinner when spinning has not yet started.

If you apply the following patch, you can see the test output show the actual spin values, which are now lower or higher than the expected asserts in the test (so the tests should be updated to match the new rates, which will cover the "new test requirement" as I mentioned):

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
index 24e69703a6..848e4289a8 100644
--- a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
+++ b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
@@ -3,8 +3,10 @@
 
 using System;
 using System.Collections.Generic;
+using System.Globalization;
 using System.Linq;
 using NUnit.Framework;
+using osu.Framework.Logging;
 using osu.Framework.Testing;
 using osu.Framework.Utils;
 using osu.Game.Beatmaps;
@@ -48,7 +50,13 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
                 PassCondition = () =>
                 {
                     var counter = Player.ChildrenOfType<SpinnerSpmCalculator>().SingleOrDefault();
-                    return counter != null && Precision.AlmostEquals(counter.Result.Value, 286, 1);
+
+                    if (counter == null)
+                        return false;
+
+                    Logger.Log(counter.Result.Value.ToString(CultureInfo.InvariantCulture));
+
+                    return Precision.AlmostEquals(counter.Result.Value, 286, 1);
                 }
             });
         }

```

![JetBrains Rider-EAP 2022-02-14 at 07 06 42](https://user-images.githubusercontent.com/191335/153816082-ba3460ac-ce45-470a-b7cd-4a889b3dc378.png)

sh0ckR6 committed:
Fix Spun Out tests

Change 'unaffected by mods' test to use dynamic RPM value instead of a fixed value

sh0ckR6 committed:
Add test to ensure spinners only complete

No bonus or a non-300 judgement

sh0ckR6 commented:
Updated the PR to include a fixed version of the `TestSpinRateUnaffectedByMods` test that now uses the dynamic RPM. I've also added a new test as mentioned above to ensure that the spinner only completes without any bonuses or a non-300 judgement.

peppy commented on osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs:
```diff
@@ -48,11 +49,43 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
+                        return false;
+
+                    // ignore cases where the spinner hasn't started as these lead to false-positives
+                    if (Precision.AlmostEquals(counter.Result.Value, 0, 1))
```
Not really sure this is required. The condition is an "until" one, so this is implicit. Also if you look at the output of tests, there's still a spin-up period where the values are non-zero but also don't match the expected rate yet, which would fall through to the check below anyway.

sh0ckR6 replied:
Yeah, it makes sense to remove this. I had written this while figuring out the calculations when my two values would sit at 0, and I actually needed this case while working on it or I would get nowhere.

sh0ckR6 commented on osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs:
```diff
@@ -48,11 +49,43 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
         }
 
+        [Test]
+        public void TestSpinnerOnlyComplete() => CreateModTest(new ModTestData
```
I was hoping you or someone else would come up with a better name for the test, it just wasn't sitting right with me.

sh0ckR6 commented on osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs:
```diff
@@ -48,11 +49,43 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
+                if (!Precision.AlmostEquals(gameplayClockContainer.CurrentTime, spinner.HitObject.StartTime + spinner.HitObject.Duration, 200.0f))
+                    return false;
+
+                return Precision.AlmostEquals(spinner.Progress, 1.0f, 0.05f) && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
```
This feels much cleaner, however I'm noticing an issue.

I've updated the test to use `spinner.IsHit` which makes more sense than using the gameplay clock. I tried looking for something like `spinner.IsCompleted`, however this is the closest thing I could find. The problem is that the spinner unloads sometimes before we can get a judgement reading, so the test fails in those cases.

With the following patch applied (which includes the updated test), the log output shows the following on a "failing" test:
```patch
diff --git a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
index e71377a505..d64f483589 100644
--- a/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
+++ b/osu.Game.Rulesets.Osu.Tests/Mods/TestSceneOsuModSpunOut.cs
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using NUnit.Framework;
+using osu.Framework.Logging;
 using osu.Framework.Testing;
 using osu.Framework.Utils;
 using osu.Game.Beatmaps;
@@ -14,7 +15,6 @@
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Osu.Objects.Drawables;
 using osu.Game.Rulesets.Osu.Skinning.Default;
-using osu.Game.Screens.Play;
 using osuTK;
 
 namespace osu.Game.Rulesets.Osu.Tests.Mods
@@ -66,7 +62,7 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
         }
 
         [Test]
-        public void TestSpinnerOnlyComplete() => CreateModTest(new ModTestData
+        public void TestSpinnerGetsNoBonusScore() => CreateModTest(new ModTestData
         {
             Mod = new OsuModSpunOut(),
             Autoplay = false,
@@ -74,15 +70,18 @@ public void TestSpinRateUnaffectedByMods(Type additionalModType)
             PassCondition = () =>
             {
                 var spinner = Player.ChildrenOfType<DrawableSpinner>().SingleOrDefault();
-                var gameplayClockContainer = Player.ChildrenOfType<GameplayClockContainer>().SingleOrDefault();
 
-                if (spinner == null || gameplayClockContainer == null)
+                if (spinner == null)
+                {
+                    Logger.Log("Spinner is null!");
                     return false;
+                }
 
-                if (!Precision.AlmostEquals(gameplayClockContainer.CurrentTime, spinner.HitObject.StartTime + spinner.HitObject.Duration, 200.0f))
-                    return false;
+                Logger.Log(spinner.IsHit.ToString());
+                Logger.Log(Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1).ToString());
+                Logger.Log("------------------------------");
 
-                return Precision.AlmostEquals(spinner.Progress, 1.0f, 0.05f) && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
+                return spinner.IsHit && Precision.AlmostEquals(spinner.GainedBonus.Value, 0, 1);
             }
         });
```
![image](https://user-images.githubusercontent.com/68568464/154806182-bcd76418-c759-46b5-a22a-f77d41f3117b.png)

Out of 50 tests, this happens ~15% of the time.
In the visual test, the spinner completes every time, but the test still fails sometimes. If there's any way to get around this race condition, please let me know because I would like to get this test passing 100% of the time.
Also as a side note, I tested the old version of this test against the `master` branch and it works every time (test fails as there are bonus spins).

peppy replied:
The remaining issue that you were noticing is from the spinner potentially going away before the next check runs. So really you needed to keep a reference to the spinner and keep checking it directly. This actually goes one level deeper, as the `DrawableSpinner` may have been "unpooled", losing its judgement state and `HitObject` reference.

I've rewritten the test to take this into account, please take a look at the changes to understand how you can write these tests in the future.

Looking for a third party review to ensure the logic I've added is correctly.

bdach replied:
I think it looks OK, but [the latest test failures](https://github.com/ppy/osu/runs/5283951075) may indicate that this same method needs to be transplanted to `TestSpinnerAutoCompleted` or others.

peppy replied:
For sure. But I guess this was a failure that has been occurring until now, because nothing in this PR should have affected this...

peppy replied:
I've fixed that test, but left the remaining one as-is (doesn't seem to be failing due to checking earlier into the spinner... and is probably not possible to write in a better way without further consideration as it's checking the spinning speed specifically).

peppy committed:
Merge branch 'master' into issues/16839-spun-out-rate

peppy committed:
Fix test not always checking the final bonus value

Due to the previous logic not waiting until the spinner had completed,
there could be false negatives as the check runs too early, with a
potential additional bonus spin occurring afterwards.

peppy committed:
Remove checks which are still going to occasionally fail due to pooling

peppy committed:
Fix completely incorrect judgement specification

peppy committed:
Replace LINQ queries with recommendations

peppy committed:
Fix intermittent failures in remaining test method

peppy committed:
Merge branch 'master' into issues/16839-spun-out-rate
